### PR TITLE
Fix for setting wrong version in CSRs.

### PR DESCRIPTION
**Pull request by anhu**
Fixes ZD 18873

# Testing
```
$ cd wolfssl
$ ./configure --enable-wolfclu
$ make all check
$ sudo make install
$cd ../wolfCLU
$ ./configure (all tried with CFLAG=-DNO_WOLFSSL_REQ_PRINT) 
$ make all
$ ./wolfssl -genkey rsa -size 2048 -out temp.key -outform pem  -output KEYPAIR
$ ./wolfssl req -new -days 1095 -config csr.conf -key temp.key.priv -out service.csr  -outform PEM -sha256
$ openssl req -in service.csr -text
        Version: 1 (0x0)
$ ./wolfssl req -in service.csr  -noout -text
        Version: 1 (0x0)
```

### CODE REVIEW

File: src/x509.c [hunk 1]
```diff
@@ -14840,6 +14842,23 @@ void wolfSSL_X509_REQ_free(WOLFSSL_X509* req)
     wolfSSL_X509_free(req);
 }
 
+int wolfSSL_X509_REQ_set_version(WOLFSSL_X509 *x, long version) {
```
Comment (dgarske) [inline src/x509.c:14845]: NIT: Functions use brace on new line.

File: src/x509.c [hunk 2]
```diff
@@ -14840,6 +14842,23 @@ void wolfSSL_X509_REQ_free(WOLFSSL_X509* req)
...
+        return WOLFSSL_FAILURE;
+    }
+    x->version = (int)version;
+    return WOLFSSL_SUCCESS;
+}
+
+long wolfSSL_X509_REQ_get_version(const WOLFSSL_X509 *req) {
+    WOLFSSL_ENTER("wolfSSL_X509_REQ_get_version");
+    if (req == NULL) {
+        return WOLFSSL_FAILURE;
```
Comment (dgarske) [inline src/x509.c:14857]: Would prefer return `return 0; /* invalid arg */` here instead of `WOLFSSL_FAILURE = 0`.

**Review by dgarske — CHANGES_REQUESTED** [review]
Looks good otherwise

File: src/x509.c [hunk 3]
```diff
@@ -14840,6 +14842,25 @@ void wolfSSL_X509_REQ_free(WOLFSSL_X509* req)
     wolfSSL_X509_free(req);
 }
 
+int wolfSSL_X509_REQ_set_version(WOLFSSL_X509 *x, long version)
```
Comment (dgarske) [inline src/x509.c:14845]: See `wolfSSL_X509_set_version` -> `x509->version = (int) v + 1;`

File: src/x509.c [hunk 4]
```diff
@@ -7067,8 +7067,10 @@ int wolfSSL_X509_REQ_print(WOLFSSL_BIO* bio, WOLFSSL_X509* x509)
             return WOLFSSL_FAILURE;
     }
 
-    /* print version of cert */
-    if (X509PrintVersion(bio, wolfSSL_X509_version(x509), 8)
+    /* print version of cert.  Note that we increment by 1 because for REQs,
+     * the value stored in x509->version is the actual value of the field; not
+     * the version. */
+    if (X509PrintVersion(bio, (int)wolfSSL_X509_REQ_get_version(x509) + 1, 8)
```
Comment (dgarske) [inline src/x509.c:7073]: Is this really the right way/place to adjust the version? The `wolfSSL_X509_get_version` and `wolfSSL_X509_set_version` do it differently.

**Review by dgarske — CHANGES_REQUESTED** [review]

**Comment by bandi13** [comment]
retest this please

File: src/x509.c (hunk 4 above)
Comment (anhu) [reply src/x509.c:7073]: Yes, I think the reason this is correct is because this is specific for REQs; not certificates.

Comment (dgarske) [reply src/x509.c:7073]: Shouldn’t the +1 be inside the get function? Or is this the same way openssl does it?

Comment (anhu) [reply src/x509.c:7073]: If the +1 was in the wolfSSL_X509_REQ_get_version() then it would return the wrong value.  This behaviour matches OpenSSL.

**Review by dgarske — APPROVED** [review]

**Pull request merged**